### PR TITLE
Add macOS 12 image templates

### PR DESCRIPTION
**Pull request by dmitry-shibanov**
# Description
In scope of this pull request we are preparing virtual-environments source code to support macOS 12 (Monterey).

[Build macOS 12](https://github.visualstudio.com/virtual-environments/_build/results?buildId=117005&view=results)
## Check list
- [ ] Related issue / work item is attached
- [x] Tests are written (if applicable)
- [x] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/macos/provision/core/xcode.ps1 [hunk 1]
```diff
@@ -15,6 +15,9 @@ if ([string]::IsNullOrEmpty($env:XCODE_INSTALL_USER) -or [string]::IsNullOrEmpty
 $env:SPACESHIP_SKIP_2FA_UPGRADE = 1
 
 $os = Get-OSVersion
+# # Monterey needs future review:
+# For now we install only Xcode 13.0
+# https://github.visualstudio.com/virtual-environments/_build/results?buildId=115815&view=results
```
Comment (maxim-lobanov) [inline images/macos/provision/core/xcode.ps1:20]: Please remove this comment completely. It is expected that Monterey doesn't support old Xcode versions and no other actions are required here.

File: images/macos/software-report/SoftwareReport.Generator.ps1 [hunk 2]
```diff
@@ -77,13 +82,21 @@ $packageManagementList = @(
...
+    (Get-NuGetVersion)
+)
+
+if ($os.IsLessThanMonterey) {
+    $packageManagementList += @(
+        (Get-CondaVersion)
+    )
+}
+
+$packageManagementList += @(
```
Comment (maxim-lobanov) [inline images/macos/software-report/SoftwareReport.Generator.ps1:94]: Why do we need these lines? Can we just put these 2 items to the `$packageManagementList` list above?

File: images/macos/tests/Node.Tests.ps1 [hunk 3]
```diff
@@ -51,7 +51,7 @@ Describe "nvm" {
     }
 }
 
-Describe "Global NPM Packages" {
+Describe "Global NPM Packages" -Skip:($os.IsMonterey) {
```
Comment (maxim-lobanov) [inline images/macos/tests/Node.Tests.ps1:54]: Why do we skip these tests?

File: images/macos/toolsets/toolset-12.json [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    "brew": {
+        "common_packages": [
+            "ant",
+            "aria2",
+            "bazelisk",
+            "carthage",
+            "cmake",
+            "gh",
+            "gnu-tar",
+            "go@1.17",
```
Comment (maxim-lobanov) [inline images/macos/toolsets/toolset-12.json:72]: Why `go@1.17` is here? Looks like incorrect merge?

File: images/macos/provision/core/ruby.sh [hunk 5]
```diff
@@ -4,6 +4,10 @@ source ~/utils/utils.sh
 DEFAULT_RUBY_VERSION=$(get_toolset_value '.ruby.default')
 echo Installing Ruby...
 brew_smart_install "ruby@${DEFAULT_RUBY_VERSION}"
+export PATH=/usr/local/opt/ruby@${DEFAULT_RUBY_VERSION}/bin:$PATH
```
Comment (maxim-lobanov) [inline images/macos/provision/core/ruby.sh:7]: I think `export` is not needed here.

File: images/macos/provision/core/ruby.sh [hunk 6]
```diff
@@ -4,6 +4,10 @@ source ~/utils/utils.sh
...
+GEM_PATH=`gem env|awk '/EXECUTABLE DIRECTORY/ {print $4}'`
+echo "GEM_PATH=$GEM_PATH" >> "$HOME/.bashrc"
+echo 'export PATH="$GEM_PATH:/usr/local/opt/ruby@3.0/bin:$PATH"'  >> "$HOME/.bashrc"
```
Comment (maxim-lobanov) [inline images/macos/provision/core/ruby.sh:10]: Should be `ruby@${DEFAULT_RUBY_VERSION}` here?

File: images/macos/provision/core/xcode.ps1 (hunk 1 above)
Comment (dmitry-shibanov) [reply images/macos/provision/core/xcode.ps1]: Done. Thank you.

File: images/macos/software-report/SoftwareReport.Generator.ps1 (hunk 2 above)
Comment (dmitry-shibanov) [reply images/macos/software-report/SoftwareReport.Generator.ps1]: Done. Thank you.

File: images/macos/helpers/Common.Helpers.psm1 [hunk 7]
```diff
@@ -28,16 +28,22 @@ function Get-EnvironmentVariable($variable) {
 function Get-OSVersion {
     $osVersion = [Environment]::OSVersion
     $osVersionMajorMinor = $osVersion.Version.ToString(2)
+    # Monterey needs future review:
+    # [Environment]::OSVersion returns 11.0 for Monterey preview.
+    $osMontereyVersion = [SemVer]$(sw_vers -productVersion)
```
Comment (al-cheb) [inline images/macos/helpers/Common.Helpers.psm1:33]: `$osMontereyVersion = [SemVer]$(sw_vers -productVersion)` -> `[SemVer]$osMontereyVersion = sw_vers -productVersion`

File: images/macos/helpers/Common.Helpers.psm1 [hunk 8]
```diff
@@ -28,16 +28,22 @@ function Get-EnvironmentVariable($variable) {
...
     return [PSCustomObject]@{
         Version = $osVersion.Version
         Platform = $osVersion.Platform
         IsHighSierra = $osVersionMajorMinor -eq "10.13"
         IsMojave = $osVersionMajorMinor -eq "10.14"
         IsCatalina = $osVersionMajorMinor -eq "10.15"
         IsBigSur = $osVersionMajorMinor -eq "11.0"
+        IsMonterey = "$($osMontereyVersion.Major)" -eq "12"
```
Comment (al-cheb) [inline images/macos/helpers/Common.Helpers.psm1:41]: `"$($osMontereyVersion.Major)"` -> `$osMontereyVersion.Major`

File: images/macos/helpers/Common.Helpers.psm1 (hunk 7 above)
Comment (dmitry-shibanov) [reply images/macos/helpers/Common.Helpers.psm1]: Done. Thank you.

File: images/macos/provision/core/ruby.sh (hunk 6 above)
Comment (dmitry-shibanov) [reply images/macos/provision/core/ruby.sh]: Done. Thank you.

File: images/macos/templates/macOS-12.json [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            "execute_command": "chmod +x {{ .Path }}; sudo {{ .Vars }} {{ .Path }}",
+            "script": "./provision/core/reboot.sh",
+            "expect_disconnect": true
+        },
+        {
+            "type": "shell",
+            "execute_command": "chmod +x {{ .Path }}; {{ .Vars }} {{ .Path }}",
+            "scripts": [
+                "./provision/core/commonutils.sh",
+                "./provision/core/golang.sh",
```
Comment (MaksimZhukov) [inline images/macos/templates/macOS-12.json:168]: It looks like, we missed the [safari.sh](https://github.com/actions/virtual-environments/blob/main/images/macos/provision/core/safari.sh) script in this section

File: images/macos/software-report/SoftwareReport.Generator.ps1 [hunk 10]
```diff
@@ -49,10 +49,15 @@ $languageAndRuntimeList = @(
 
 if ( -not $os.IsHighSierra) {
     $languageAndRuntimeList += @(
-        (Get-GccVersion)
         (Get-FortranVersion)
```
Comment (MaksimZhukov) [inline images/macos/software-report/SoftwareReport.Generator.ps1:52]: It looks like, we also need to add a condition for the `Get-NewmanVersion` and `Get-FortranVersion` to skip them for Monterey

**Review by MaksimZhukov — APPROVED** [review]

File: images/macos/software-report/SoftwareReport.Generator.ps1 (hunk 10 above)
Comment (dmitry-shibanov) [reply images/macos/software-report/SoftwareReport.Generator.ps1]: Done. Thank you.

File: images/macos/templates/macOS-12.json (hunk 9 above)
Comment (dmitry-shibanov) [reply images/macos/templates/macOS-12.json:168]: Done. Thank you.

File: images/macos/tests/Node.Tests.ps1 (hunk 3 above)
Comment (dmitry-shibanov) [reply images/macos/tests/Node.Tests.ps1]: Done. Thank you.

File: images/macos/toolsets/toolset-12.json (hunk 4 above)
Comment (dmitry-shibanov) [reply images/macos/toolsets/toolset-12.json]: Done. Thank you.

File: images/macos/provision/core/ruby.sh (hunk 5 above)
Comment (dmitry-shibanov) [reply images/macos/provision/core/ruby.sh:7]: Done. Thank you.

**Review by MaksimZhukov — APPROVED** [review]

**Review by al-cheb — APPROVED** [review]

**Review by miketimofeev — DISMISSED** [review]

**Review by dibir-magomedsaygitov — DISMISSED** [review]

File: docs/software-and-images-guidelines.md [hunk 11]
```diff
@@ -31,7 +31,7 @@ These are the guidelines we follow in software and images supporting routine:
 | PyPy      | 3 most popular `major.minor` versions |
 | .NET Core | 2 latest LTS versions and 1 latest version. For each feature version only latest patch is installed |
 | GCC <br/> GNU Fortran <br/> Clang <br/> GNU C++ | 3 latest major versions |
-| Android NDK | 1 latest, 1 LTS version |
+| Android NDK | 1 latest non-LTS, 2 latest LTS versions |
```
Comment (miketimofeev) [inline docs/software-and-images-guidelines.md:34]: Looks like the merge went wrong

**Review by miketimofeev — DISMISSED** [review]

**Review by miketimofeev — APPROVED** [review]

**Review by MaksimZhukov — APPROVED** [review]

**Review by miketimofeev — APPROVED** [review]

**Pull request merged**